### DISCUSSION TITLE
Error when build c/c++ dependencies: can't found

### DISCUSSION THREAD

**Discussion by zooeywm** [rust-cross/cargo-zigbuild#246] [Q&A] [answered]
```
  cargo:warning=./src/handlers/my_handler.hpp:14:10: fatal error: 'X11/Xatom.h' file not found

  cargo:warning=#include <X11/Xatom.h>

  cargo:warning=         ^~~~~~~~~~~~~
```

I'm a C/C++ novice, I have a rust project that has C/C++ dependencies, X11/Xatom.h This library can be found during normal cargo build —target x86_64-unknown-linux-gnu compilation, when using zigbuild, the X11/Xatom.h` cannot be found when specifying any version of glibc on the same platform. Thanks for any help!

**Comment by polarathene** [comment, marked as answer]
Assuming that `X11/Xatom.h` is found at `/usr/include`, this should be fixable by including the system directory (_which `zig` excludes when `-nostdinc` / `-target` is configured which `cargo zigbuild` sets_).

This should work for you (_`RUSTFLAGS` may not be necessary sometimes_):

```bash
CFLAGS='-isystem /usr/include' RUSTFLAGS='-L /usr/lib64' cargo zigbuild --release --target x86_64-unknown-linux-gnu
```

Zig should provide it's own headers and those should be preferred when building for older glibc. You would not want to use `-I /usr/include` which would introduce conflicts. This should provide fallback to `/usr/include` for additional header files from other system packages.

## Full explanation

The default system libs like `/usr/include` are excluded when you use zig with `-target`, but in the case of `cargo-zigbuild` it always enforces this regardless due to adding `-nostdinc`:

https://github.com/rust-cross/cargo-zigbuild/blob/eba2d7e6691ebbe59b8ecf4726cef4f49ec0d1b4/src/zig.rs#L839-L841

To include that again you can use args like `-isystem` / `-idirafter` with `/usr/include`:

```bash
# `cargo-zigbuild` effectively runs `zig cc -nostdinc`,
# which excludes system include dir `/usr/include`, add it back via `-isystem`:
CC='zig cc -nostdinc -isystem /usr/include' make install
```

When you use `-target` with `zig cc`, such as `-target x86_64-linux-gnu` or the custom glibc target (eg: `-target x86_64-linux-gnu.2.34`), that will also exclude default search dirs for linking shared libraries (`.so` files), so you'll need to explicitly provide those too such as with `-L /usr/lib64` (_to `zig cc` like above with `-isystem`, or `RUSTFLAGS`_).

---

These two extra flags need to be picked up separately usually if you've got a common command like `cargo-zigbuild` (_which implicitly configures both `CC` and `RUSTFLAGS='-C linker=...'` for you_).
- For `zig cc` there is the `CFLAGS` ENV, and for `zig c++` you'd have `CXXFLAGS` equivalent as a way to pass extra options to the `CC` / `CXX` ENV or associated commands. Those same standard ENV work with `gcc`/`g++` and `clang`. In Rust this can be relevant for build-time dependencies like `-sys` crates with their `build.rs`.
- Next is linking libs by the linker (`RUSTFLAGS -C linker=...`). `zig` / `gcc` / `clang` can leverage the `LDFLAGS` ENV for providing extra linker args, but for `rustc` you need to provide this information via `RUSTFLAGS` ENV so when crates have a `build.rs` that explicitly link libs via `rustc-link-lib` that it'll know about extra search directories.

Hence here's how `zig cc` could be used with a typical project that has a `Makefile` vs the equivalent for Rust via `cargo`:

```bash
CC='zig cc -target x86_64-linux-gnu' CFLAGS='-isystem /usr/include' LDFLAGS='-L /usr/lib64' make install

CFLAGS='-isystem /usr/include' RUSTFLAGS='-L /usr/lib64' cargo zigbuild --release --target x86_64-unknown-linux-gnu
```
